engine/src/main/java/com/odysseusinc/arachne/executionengine/service/impl/DockerServiceImpl.java

```diff
@@ -0,0 +1,202 @@
+                .withRemotePath("/code").exec();
+
+        String baseDir = "/code/" + analysisDir.getName() + "/";
+        String workingDir = baseDir + (execFileName.split("/").length > 1 ? execFileName.split("/")[0] : "");
+        LOGGER.info("Working dir {}", workingDir);
+        String rscriptPath = baseDir + execFileName;
+        LOGGER.info("Script path {}", rscriptPath);
+        String runCommandId = dockerClient.execCreateCmd(containerId)
+                .withAttachStdout(true)
+                .withAttachStderr(true)
+                .withCmd("Rscript", rscriptPath)
+                .withWorkingDir(workingDir).exec()
+                .getId();
+
+        ResultCallback.Adapter<Frame> adapter = getAdapter(analysisDir, analysis, callback);
+        try {
+            LOGGER.info("Running script: {}", rscriptPath);
+            dockerClient.execStartCmd(runCommandId).exec(adapter).awaitCompletion();
+        } catch (InterruptedException e) {
+            LOGGER.error("InterruptedException occurred whilst running {} in {}, message: {}", rscriptPath, containerId, e.getMessage());
+            callback.execute(AnalysisResultStatusDTO.FAILED, null, analysisDir, e);
+        } finally {
+            stop(dockerClient, containerId);
+        }
+    }
+
+    private ResultCallback.Adapter<Frame> getAdapter(File analysisDir, AnalysisSyncRequestDTO analysis, AnalysisCallback callback) {
+        return new ResultCallback.Adapter<Frame>() {
+            @Override
+            public void onNext(Frame item) {
+                LOGGER.info(item.toString(), analysis);
+                try {
+                    String output = new String(item.getPayload());
+                    LOGGER.info("Execution output: {}", output);
+                    callback.execute(AnalysisResultStatusDTO.EXECUTED, output, analysisDir, null);
+                } catch (Exception e) {
+                    LOGGER.error("Error processing execution output", e);
+                    callback.execute(AnalysisResultStatusDTO.FAILED, null, analysisDir, e);
+                }
+            }
+
+            @Override
+            public void onError(Throwable throwable) {
+                LOGGER.error("Error during execution", throwable);
+                try {
+                    String error = throwable.getMessage();
+                    LOGGER.error("Execution error: {}", error);
+                    callback.execute(AnalysisResultStatusDTO.FAILED, error, analysisDir, null);
+                } catch (Exception e) {
+                    LOGGER.error("Error processing execution error", e);
+                    callback.execute(AnalysisResultStatusDTO.FAILED, null, analysisDir, e);
+                }
+            }
+
+            @Override
+            public void onComplete() {
+                LOGGER.info("Execution completed");
+            }
+
+            @Override
+            public void close() throws IOException {
+                super.close();
```

Comment:
Why do you need to override here if you mere call super?

Reply:
In this method of the ResultCallback.Adapter<Frame>, we typically close any resources that need to be explicitly released or cleaned up. At this time, there are no explicit resources that need to be closed yet but I believe it would be good leave it as a placeholder and calling the close method of the superclass (ResultCallback.Adapter<Frame>) to handle any potential cleanup operations defined in the base class and we have also the flexibiity to add our own cleanup logic if necessary in case would be needed.

Reply:
There is flexibility to add the method later on. That will not cause any collateral. I still don't see a reason to put in this override now.